File: packages/templates/src/research-agent/templates/langchain-python/tools/web_search.py
"""
Web Search Tool

Searches the web for information using a search API.

To use with real search:
1. Sign up for a search API (Brave Search, SerpAPI, or Tavily)
2. Add your API key to .env
3. Uncomment the appropriate implementation below
"""

from langchain.tools import Tool
from typing import List, Dict
import os


def web_search(query: str, max_results: int = 5) -> List[Dict[str, str]]:
    """
    Search the web for information.

    Args:
        query: Search query
        max_results: Maximum number of results

    Returns:
        List of search results with title, url, and snippet
    """

    # ================================================
    # OPTION 1: Brave Search API (Recommended)
    # ================================================
    # Uncomment to use Brave Search:
    #
    # import requests
    # api_key = os.getenv("BRAVE_API_KEY")
    # url = "https://api.search.brave.com/res/v1/web/search"
    # headers = {"X-Subscription-Token": api_key}
    # params = {"q": query, "count": max_results}
    #
    # response = requests.get(url, headers=headers, params=params)
    # data = response.json()
    #
    # results = []
    # for item in data.get("web", {}).get("results", []):
    #     results.append({
    #         "title": item.get("title", ""),
    #         "url": item.get("url", ""),
    #         "snippet": item.get("description", "")
    #     })
    # return results

    # ================================================
    # OPTION 2: SerpAPI (Google Search)
    # ================================================
    # Uncomment to use SerpAPI:
    #
    # from serpapi import GoogleSearch
    # api_key = os.getenv("SERPAPI_KEY")
    # params = {
    #     "q": query,
    #     "api_key": api_key,
    #     "num": max_results
    # }
    #
    # search = GoogleSearch(params)
    # data = search.get_dict()
    #
    # results = []
    # for item in data.get("organic_results", []):
    #     results.append({
    #         "title": item.get("title", ""),
    #         "url": item.get("link", ""),
    #         "snippet": item.get("snippet", "")
    #     })
    # return results

    # ================================================
    # OPTION 3: Tavily Search (AI-optimized)
    # ================================================
    # Uncomment to use Tavily:
    #
    # from tavily import TavilyClient
    # api_key = os.getenv("TAVILY_API_KEY")
    # client = TavilyClient(api_key=api_key)
    #
    # response = client.search(query, max_results=max_results)
    #
    # results = []
    # for item in response.get("results", []):
    #     results.append({
    #         "title": item.get("title", ""),
    #         "url": item.get("url", ""),
    #         "snippet": item.get("content", "")
    #     })
    # return results

    # ================================================
    # PLACEHOLDER (for testing without API)
    # ================================================
    print(f"[MOCK] Searching for: {query}")
    return [
        {
            "title": f"Search result for: {query}",
            "url": f"https://example.com/search?q={query.replace(' ', '+')}",
            "snippet": f"This is a mock search result for '{query}'. "
                      f"Set up a real search API to get actual results. "
                      f"See web_search.py for integration options."
        }
    ]


# Create LangChain tool
web_search_tool = Tool(
    name="web_search",
    description=(
        "Search the web for information on any topic. "
        "Useful for finding current information, facts, news, and articles. "
        "Input should be a clear, specific search query. "
        "Returns a list of relevant web pages with titles and snippets."
    ),
    func=lambda query: web_search(query, max_results=5)
)

File: packages/agents/src/tools/web-search.ts
import axios from "axios";
import { AgentTool } from "../types";

/**
 * Web search tool using Brave Search API.
 *
 * Requires BRAVE_SEARCH_API_KEY environment variable,
 * or falls back to basic web scraping of search results.
 */
export const webSearchTool: AgentTool = {
  name: "web_search",
  description:
    "Search the web for information. Returns a list of results with titles, URLs, and snippets. Use this to find current information, research topics, or verify facts.",
  input_schema: {
    type: "object" as const,
    properties: {
      query: {
        type: "string",
        description: "The search query",
      },
      max_results: {
        type: "number",
        description: "Maximum number of results to return (default: 5)",
      },
    },
    required: ["query"],
  },
  async execute(input: Record<string, unknown>): Promise<string> {
    const query = input.query as string;
    const maxResults = (input.max_results as number) ?? 5;

    const braveKey = process.env.BRAVE_SEARCH_API_KEY;

    if (braveKey) {
      return searchWithBrave(query, maxResults, braveKey);
    }

    // Fallback: use DuckDuckGo instant answer API (limited but free, no key needed)
    return searchWithDDG(query, maxResults);
  },
};

async function searchWithBrave(
  query: string,
  maxResults: number,
  apiKey: string
): Promise<string> {
  const response = await axios.get("https://api.search.brave.com/res/v1/web/search", {
    params: { q: query, count: maxResults },
    headers: {
      Accept: "application/json",
      "Accept-Encoding": "gzip",
      "X-Subscription-Token": apiKey,
    },
    timeout: 10000,
  });

  const results = response.data.web?.results ?? [];

  if (results.length === 0) {
    return `No results found for "${query}".`;
  }

  const formatted = results
    .slice(0, maxResults)
    .map(
      (r: { title: string; url: string; description: string }, i: number) =>
        `${i + 1}. **${r.title}**\n   URL: ${r.url}\n   ${r.description}`
    )
    .join("\n\n");

  return `Search results for "${query}":\n\n${formatted}`;
}

async function searchWithDDG(query: string, maxResults: number): Promise<string> {
  try {
    const response = await axios.get("https://api.duckduckgo.com/", {
      params: {
        q: query,
        format: "json",
        no_redirect: 1,
        no_html: 1,
      },
      timeout: 10000,
    });

    const data = response.data;
    const results: string[] = [];

    // Abstract (main answer)
    if (data.Abstract) {
      results.push(`**Summary**: ${data.Abstract}\nSource: ${data.AbstractURL}`);
    }

    // Related topics
    if (data.RelatedTopics) {
      for (const topic of data.RelatedTopics.slice(0, maxResults)) {
        if (topic.Text && topic.FirstURL) {
          results.push(`- ${topic.Text}\n  URL: ${topic.FirstURL}`);
        }
      }
    }

    if (results.length === 0) {
      return `No results found for "${query}". Try a more specific search query or ensure BRAVE_SEARCH_API_KEY is set for better results.`;
    }

    return `Search results for "${query}":\n\n${results.join("\n\n")}`;
  } catch {
    return `Search failed for "${query}". Set BRAVE_SEARCH_API_KEY for reliable search results.`;
  }
}
